File: src/python/_old_flat/class_binding_handler.py
"""
Class Binding Handler - Xử lý @class directive
Supports: static classes, conditional classes, và mixed binding
Uses separate __classBinding() method for reactive class management
"""

import re
from php_js_converter import php_to_js_advanced

class ClassBindingHandler:
    def __init__(self, state_variables=None):
        self.state_variables = state_variables or set()
    
    def process_class_directive(self, content):
        """
        Process @class directive with multiple formats:
        1. @class('static-class') - static class
        2. @class('class', $condition) - conditional class  
        3. @class(['class1' => $cond1, 'class2' => $cond2]) - array binding
        4. @class(['static', 'dynamic' => $cond]) - mixed
        
        Multiple @class on same line will be merged into one
        
        Output format:
        - Static only: class="static-class"
        - With dynamic: ${this.__classBinding([...])}
        
        Note: Supports multi-line directives within HTML tag attributes
        """
        result = content
        
        # Process all @class directives globally (supports multi-line)
        while True:
            match = re.search(r'@class\s*\(', result)
            if not match:
                break
            
            # Find matching closing parenthesis (can span multiple lines)
            start_pos = match.end() - 1
            paren_count = 0
            i = start_pos
            
            while i < len(result):
                if result[i] == '(':
                    paren_count += 1
                elif result[i] == ')':
                    paren_count -= 1
                    if paren_count == 0:
                        # Found complete directive
                        expression = result[start_pos + 1:i].strip()
                        replacement = self._generate_class_output(expression)
                        result = result[:match.start()] + replacement + result[i + 1:]
                        break
                i += 1
            else:
                # No matching closing paren found - skip this one
                break
        
        return result
    
    def _merge_multiple_class_directives(self, line):
        """
        Merge multiple @class directives on the same line into one
        """
        all_bindings = []
        
        # Extract all @class expressions
        while True:
            match = re.search(r'@class\s*\(', line)
            if not match:
                break
            
            start_pos = match.end() - 1
            paren_count = 0
            i = start_pos
            
            # Find matching closing parenthesis
            while i < len(line):
                if line[i] == '(':
                    paren_count += 1
                elif line[i] == ')':
                    paren_count -= 1
                    if paren_count == 0:
                        expression = line[start_pos + 1:i].strip()
                        bindings = self._parse_class_expression(expression)
                        all_bindings.extend(bindings)
                        # Remove this @class directive
                        line = line[:match.start()] + line[i + 1:]
                        break
                i += 1
            else:
                break
        
        # Generate single output for all merged bindings
        if all_bindings:
            replacement = self._generate_output_from_bindings(all_bindings)
            # Insert at the first @class position (which is now gone)
            # Find a good insertion point - before the first > or at the start
            insertion_point = line.find('>')
            if insertion_point > 0:
                line = line[:insertion_point] + ' ' + replacement + line[insertion_point:]
            else:
                line = line + ' ' + replacement
        
        return line
    
    def _process_single_class_directive(self, line):
        """
        Process a single @class directive on a line
        """
        match = re.search(r'@class\s*\(', line)
        if not match:
            return line
        
        start_pos = match.end() - 1
        paren_count = 0
        i = start_pos
        
        while i < len(line):
            if line[i] == '(':
                paren_count += 1
            elif line[i] == ')':
                paren_count -= 1
                if paren_count == 0:
                    expression = line[start_pos + 1:i].strip()
                    replacement = self._generate_class_output(expression)
                    line = line[:match.start()] + replacement + line[i + 1:]
                    break
            i += 1
        
        return line
    
    def _generate_class_output(self, expression):
        """
        Generate output for class directive
        Returns either static class="..." or ${this.__attr({"class": {...}})}
        """
        bindings = self._parse_class_expression(expression)
        return self._generate_output_from_bindings(bindings)
    
    def _generate_output_from_bindings(self, bindings):
        """
        Generate output from a list of bindings
        Returns either static class="..." or ${this.__classBinding([...])}
        """
        # Check if all bindings are static
        has_dynamic = any(b['type'] == 'binding' for b in bindings)
        
        if not has_dynamic:
            # All static - output direct class attribute
            static_classes = [b['value'] for b in bindings]
            return f'class="{" ".join(static_classes)}"'
        else:
            # Has dynamic - use __classBinding
            return self._generate_class_binding_config(bindings)
    
    def _generate_class_binding_config(self, bindings):
        """
        Generate __classBinding() config
        Format: ${this.__classBinding([{type: "static", value: "...", states: [...], checker: () => ...}])}
        """
        binding_configs = []
        
        for binding in bindings:
            if binding['type'] == 'static':
                binding_configs.append(f'{{type: "static", value: "{binding["value"]}"}}')
            elif binding['type'] == 'binding':
                states = binding.get('states', [])
                states_str = ', '.join([f'"{s}"' for s in states])
                checker = binding['checker']
                class_name = binding['value']
                binding_configs.append(f'{{type: "binding", value: "{class_name}", states: [{states_str}], checker: () => {checker}}}')
        
        configs_joined = ', '.join(binding_configs)
        return f'${{this.__classBinding([{configs_joined}])}}'
    
    def _parse_class_expression(self, expression):
        """
        Parse class expression and return list of binding dicts
        Returns: [{"type": "static"|"binding", "value": "class-name", ...}, ...]
        """
        expression = expression.strip()
        
        # Case 1: Simple string - @class('class')
        if self._is_simple_string(expression):
            class_name = self._extract_string_value(expression)
            return [{"type": "static", "value": class_name}]
        
        # Case 2: Two arguments - @class('class', $condition)
        if ',' in expression and not expression.startswith('['):
            parts = self._split_arguments(expression)
            if len(parts) == 2:
                class_name = self._extract_string_value(parts[0].strip())
                condition = parts[1].strip()
                states = self._extract_state_variables(condition)
                js_condition = self._convert_php_to_js(condition)
                return [{
                    "type": "binding",
                    "value": class_name,
                    "states": states,
                    "checker": js_condition
                }]
        
        # Case 3: Array format - @class(['class1' => $cond, 'class2'])
        if expression.startswith('[') and expression.endswith(']'):
            return self._parse_array_expression(expression)
        
        # Fallback: treat as static class
        return [{"type": "static", "value": expression}]
    
    def _parse_array_expression(self, expression):
        """
        Parse array expression: ['class1' => $cond, 'static', 'class2' => $cond2]
        Returns list of binding dicts
        """
        # Remove outer brackets
        inner = expression[1:-1].strip()
        
        # Split by comma, but respect nested structures
        items = self._split_array_items(inner)
        
        bindings = []
        for item in items:
            item = item.strip()
            if not item:
                continue
            
            # Check if it's a key => value pair
            if '=>' in item:
                parts = item.split('=>', 1)
                class_name = self._extract_string_value(parts[0].strip())
                condition = parts[1].strip()
                
                # Check if condition is a static string
                if self._is_simple_string(condition):
                    # Static value like 'demo' => 'dump'
                    static_value = self._extract_string_value(condition)
                    bindings.append({"type": "static", "value": static_value})
                else:
                    # Dynamic condition
                    states = self._extract_state_variables(condition)
                    js_condition = self._convert_php_to_js(condition)
                    bindings.append({
                        "type": "binding",
                        "value": class_name,
                        "states": states,
                        "checker": js_condition
                    })
            else:
                # Static class without condition
                class_name = self._extract_string_value(item)
                bindings.append({"type": "static", "value": class_name})
        
        return bindings
    
    def _split_array_items(self, content):
        """
        Split array items by comma, respecting nested structures
        """
        items = []
        current = []
        depth = 0
        in_string = False
        string_char = None
        i = 0
        
        while i < len(content):
            char = content[i]
            
            # Handle string boundaries
            if char in ('"', "'") and (i == 0 or content[i-1] != '\\'):
                if not in_string:
                    in_string = True
                    string_char = char
                elif char == string_char:
                    in_string = False
                    string_char = None
            
            # Handle nested structures
            if not in_string:
                if char in '([{':
                    depth += 1
                elif char in ')]}':
                    depth -= 1
                elif char == ',' and depth == 0:
                    items.append(''.join(current))
                    current = []
                    i += 1
                    continue
            
            current.append(char)
            i += 1
        
        if current:
            items.append(''.join(current))
        
        return items
    
    def _split_arguments(self, expression):
        """
        Split function arguments by comma, respecting nested structures
        """
        return self._split_array_items(expression)
    
    def _is_simple_string(self, expression):
        """
        Check if expression is a simple quoted string
        """
        expression = expression.strip()
        return (expression.startswith("'") and expression.endswith("'")) or \
               (expression.startswith('"') and expression.endswith('"'))
    
    def _extract_string_value(self, expression):
        """
        Extract string value from quoted string
        'class' -> class
        "class" -> class
        """
        expression = expression.strip()
        if (expression.startswith("'") and expression.endswith("'")) or \
           (expression.startswith('"') and expression.endswith('"')):
            return expression[1:-1]
        return expression
    
    def _extract_state_variables(self, php_expression):
        """
        Extract state variable names from PHP expression
        $abc -> abc
        $abc->test($cde) -> abc, cde
        $demo && !$on || !$r->e -> demo, on, r
        """
        # Find all $variable patterns
        variables = set()
        matches = re.findall(r'\$([a-zA-Z_][a-zA-Z0-9_]*)', php_expression)
        for var in matches:
            # Only include if it's a state variable
            if var in self.state_variables or True:  # Include all for now
                variables.add(var)
        
        return sorted(list(variables))
    
    def _convert_php_to_js(self, php_expression):
        """
        Convert PHP expression to JavaScript
        Uses the existing php_to_js_advanced function
        """
        return php_to_js_advanced(php_expression)

File: src/Directive/ClassBindingHandler.php
<?php

declare(strict_types=1);

namespace Saola\Compiler\Directive;

use Saola\Compiler\Expr\ExpressionCompiler;

/** Port của sao2js/class_binding_handler.py. */
final class ClassBindingHandler
{
    /** @param iterable<string> $stateVariables */
    public function __construct(iterable $stateVariables = [], private readonly ExpressionCompiler $expressions = new ExpressionCompiler())
    {
        // Python hiện chủ động thu mọi biến trong condition (`or True`), nên
        // tham số này chỉ được giữ để tương thích API và cho lần siết scope sau.
    }

    public function processClassDirective(string $content): string
    {
        $result = $content;
        while (preg_match('/@class\s*\(/', $result, $match, PREG_OFFSET_CAPTURE) === 1) {
            $matchText = $match[0][0];
            $matchStart = $match[0][1];
            $open = $matchStart + strlen($matchText) - 1;
            $depth = 0;
            $closed = false;
            $length = strlen($result);

            for ($i = $open; $i < $length; $i++) {
                if ($result[$i] === '(') {
                    $depth++;
                } elseif ($result[$i] === ')') {
                    $depth--;
                    if ($depth === 0) {
                        $expression = trim(substr($result, $open + 1, $i - $open - 1));
                        $replacement = $this->generateClassOutput($expression);
                        $result = substr($result, 0, $matchStart) . $replacement . substr($result, $i + 1);
                        $closed = true;
                        break;
                    }
                }
            }

            if (! $closed) {
                break;
            }
        }

        return $result;
    }

    /** @param list<array{type: string, value: string, states?: list<string>, checker?: string}> $bindings */
    private function generateOutputFromBindings(array $bindings): string
    {
        $dynamic = false;
        foreach ($bindings as $binding) {
            if ($binding['type'] === 'binding') {
                $dynamic = true;
                break;
            }
        }
        if (! $dynamic) {
            return 'class="' . implode(' ', array_column($bindings, 'value')) . '"';
        }

        $configs = [];
        foreach ($bindings as $binding) {
            if ($binding['type'] === 'static') {
                $configs[] = '{type: "static", value: "' . $binding['value'] . '"}';
                continue;
            }
            $states = implode(', ', array_map(static fn (string $state): string => '"' . $state . '"', $binding['states'] ?? []));
            $configs[] = '{type: "binding", value: "' . $binding['value'] . '", states: [' . $states . '], checker: () => ' . $binding['checker'] . '}';
        }

        return '${this.__classBinding([' . implode(', ', $configs) . '])}';
    }

    private function generateClassOutput(string $expression): string
    {
        return $this->generateOutputFromBindings($this->parseClassExpression($expression));
    }

    /** @return list<array{type: string, value: string, states?: list<string>, checker?: string}> */
    private function parseClassExpression(string $expression): array
    {
        $expression = trim($expression);
        if ($this->isSimpleString($expression)) {
            return [['type' => 'static', 'value' => $this->extractStringValue($expression)]];
        }

        if (str_contains($expression, ',') && ! str_starts_with($expression, '[') && ! str_starts_with($expression, '{')) {
            $parts = $this->splitArrayItems($expression);
            if (count($parts) === 2) {
                $condition = trim($parts[1]);
                return [[
                    'type' => 'binding',
                    'value' => $this->extractStringValue(trim($parts[0])),
                    'states' => $this->extractStateVariables($condition),
                    'checker' => $this->expressions->compile($condition),
                ]];
            }
        }

        if (
            (str_starts_with($expression, '[') && str_ends_with($expression, ']'))
            || (str_starts_with($expression, '{') && str_ends_with($expression, '}'))
        ) {
            return $this->parseArrayExpression($expression);
        }

        return [['type' => 'static', 'value' => $expression]];
    }

    /** @return list<array{type: string, value: string, states?: list<string>, checker?: string}> */
    private function parseArrayExpression(string $expression): array
    {
        $bindings = [];
        foreach ($this->splitArrayItems(trim(substr($expression, 1, -1))) as $item) {
            $item = trim($item);
            if ($item === '') {
                continue;
            }
            [$className, $condition] = $this->splitEntry($item);
            if ($condition === null) {
                $bindings[] = ['type' => 'static', 'value' => $this->extractStringValue($item)];
            } elseif ($this->isSimpleString($condition)) {
                $bindings[] = ['type' => 'static', 'value' => $this->extractStringValue($condition)];
            } else {
                $bindings[] = [
                    'type' => 'binding',
                    'value' => $className,
                    'states' => $this->extractStateVariables($condition),
                    'checker' => $this->expressions->compile($condition),
                ];
            }
        }

        return $bindings;
    }

    /** @return array{0: string, 1: string|null} */
    private function splitEntry(string $item): array
    {
        if (preg_match('~^\s*([\'\"])(.*?)\1\s*(?:=>|:(?!:))\s*(.+)$~su', $item, $match) === 1) {
            return [trim($match[2]), trim($match[3])];
        }
        if (preg_match('~^\s*([A-Za-z_][\w-]*)\s*(?:=>|:(?!:))\s*(.+)$~su', $item, $match) === 1) {
            return [trim($match[1]), trim($match[2])];
        }
        if (str_contains($item, '=>')) {
            [$key, $condition] = explode('=>', $item, 2);
            return [$this->extractStringValue(trim($key)), trim($condition)];
        }

        return [$item, null];
    }

    /** @return list<string> */
    private function splitArrayItems(string $content): array
    {
        $items = [];
        $current = '';
        $depth = 0;
        $quote = null;
        $length = strlen($content);
        for ($i = 0; $i < $length; $i++) {
            $char = $content[$i];
            if (($char === '"' || $char === "'") && ($i === 0 || $content[$i - 1] !== '\\')) {
                if ($quote === null) {
                    $quote = $char;
                } elseif ($quote === $char) {
                    $quote = null;
                }
            }
            if ($quote === null) {
                if (str_contains('([{', $char)) {
                    $depth++;
                } elseif (str_contains(')]}', $char)) {
                    $depth--;
                } elseif ($char === ',' && $depth === 0) {
                    $items[] = $current;
                    $current = '';
                    continue;
                }
            }
            $current .= $char;
        }
        if ($current !== '') {
            $items[] = $current;
        }

        return $items;
    }

    private function isSimpleString(string $expression): bool
    {
        $expression = trim($expression);
        return (str_starts_with($expression, "'") && str_ends_with($expression, "'"))
            || (str_starts_with($expression, '"') && str_ends_with($expression, '"'));
    }

    private function extractStringValue(string $expression): string
    {
        $expression = trim($expression);
        return $this->isSimpleString($expression) ? substr($expression, 1, -1) : $expression;
    }

    /** @return list<string> */
    private function extractStateVariables(string $expression): array
    {
        preg_match_all('/\$([a-zA-Z_][a-zA-Z0-9_]*)/', $expression, $matches);
        $variables = array_values(array_unique($matches[1]));
        sort($variables, SORT_STRING);

        return $variables;
    }
}
